Firestore/core/src/firebase/firestore/api/input_validation_std.cc

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/api/input_validation.h"
+
+#include <cstdlib>
+#include <stdexcept>
+
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+
+namespace firebase {
+namespace firestore {
+namespace api {
+namespace impl {
+
+namespace {
+
+template <typename T>
+[[noreturn]] void Throw(const char* kind, const T& error) {
+#ifdef ABSL_HAVE_EXCEPTIONS
+  (void)kind;
+  throw error;
```

Comment:
I'm pretty supportive of this. Having said that, I don't think we have previously given exception safety sufficient thought to make sure client code can recover after an exception is thrown. What do you think? I don't think we have any major problems, but perhaps it would be worthwhile to do an investigation and see if anything comes up.

Reply:
We're notably not exception safe, so if that's the bar I think we should avoid this. I've reverted this. We can revisit in the future if this becomes something customers want.

Note that abseil itself only uses this pattern in circumstances where they're implementing the standard and the standard requires an exception.

Reply:
I think that if we expect customers to find exceptions usable, we should provide at least the basic guarantee. We're mostly relying on well-behaved RAII classes, but without designing for that explicitly, it's quite likely we have some issues. I agree with leaving this as potential future work.